File: timetable-ui/src/app/access-level.ts
/**
 * The three access levels a permission grant can carry, mirroring the backend's
 * `org.lnu.timetable.security.AccessLevel` and the `access_level` PostgreSQL enum.
 *
 * They are *ordered*, and every permission question in the client reduces to one comparison:
 * editing needs `EDIT`, deleting needs `FULL`, handing access to somebody else needs `MANAGE`.
 * Keeping it a chain rather than a set of independent flags is what keeps the UI honest — there is
 * one dropdown to grant access, one word to read back, and no combination anybody has to reason
 * about that doesn't exist.
 */
export type AccessLevel = 'EDIT' | 'FULL' | 'MANAGE';

/** Weakest first — the order the grant form offers them in. */
export const ACCESS_LEVELS: readonly AccessLevel[] = ['EDIT', 'FULL', 'MANAGE'];

const RANK: Record<AccessLevel, number> = { EDIT: 0, FULL: 1, MANAGE: 2 };

/** Does `held` (absent means no access at all) satisfy a requirement of `required`? */
export function allows(held: AccessLevel | null | undefined, required: AccessLevel): boolean {
  return held != null && RANK[held] >= RANK[required];
}

/** The stronger of two levels; either may be absent. */
export function maxLevel(a: AccessLevel | null | undefined, b: AccessLevel | null | undefined): AccessLevel | null {
  if (!a) return b ?? null;
  if (!b) return a;
  return RANK[a] >= RANK[b] ? a : b;
}

export const ACCESS_LEVEL_LABELS: Record<AccessLevel, string> = {
  EDIT: 'Редагування',
  FULL: 'Повний доступ',
  MANAGE: 'Керування доступом'
};

/** The one-line explanation shown beside each option in the grant form. */
export const ACCESS_LEVEL_HINTS: Record<AccessLevel, string> = {
  EDIT: 'Створення та редагування цього ресурсу й усього, що йому підпорядковане. Без видалення.',
  FULL: 'Те саме, що «Редагування», а також видалення.',
  MANAGE: 'Те саме, що «Повний доступ», а також надання доступу іншим до цього ресурсу та підпорядкованих йому.'
};

File: timetable/src/main/java/org/lnu/timetable/security/AccessLevel.java
package org.lnu.timetable.security;

/**
 * How much a permission grant lets its holder do inside the scope it names. Three values, ordered
 * from least to most, mirroring the {@code access_level} PostgreSQL enum in {@code schema.sql}
 * (declaration order is comparison order there too, so {@code level >= 'FULL'} means the same thing
 * in SQL as {@link #allows(AccessLevel)} does here).
 * <p>
 * The reason this is one ordered dimension rather than a set of independent flags — "may update",
 * "may delete", "may delegate" — is that every combination anybody actually asked for is a prefix
 * of this chain. Nobody at the university wants a person who can delete a кафедра but not rename
 * it, or hand out access they do not themselves have. Keeping it ordered means every authorization
 * question in the service reduces to a single comparison, the admin UI is one dropdown instead of
 * a matrix of checkboxes, and "who can do what here" is answerable by reading one word.
 *
 * @see PermissionService
 */
public enum AccessLevel {

    /**
     * Create and update this resource and everything below it in the cascade; no deletes, no
     * delegation. The everyday level — a методист maintaining навчальні плани and навантаження who
     * must not be able to erase a група by mis-clicking.
     */
    EDIT("Редагування"),

    /**
     * Everything {@link #EDIT} allows, plus deleting. Separate from EDIT because deletion in this
     * schema cascades: removing a DegreeProgram takes its academic groups, curriculum items and
     * workloads with it, so it is a qualitatively different act from editing a field.
     */
    FULL("Повний доступ"),

    /**
     * Everything {@link #FULL} allows, plus granting and revoking access to this resource and its
     * descendants, at any level up to MANAGE itself. This is the delegation level: a деканат holding
     * MANAGE on their факультет can hand out EDIT on individual кафедри without an administrator
     * being involved, and can promote a deputy to MANAGE of a single кафедра.
     */
    MANAGE("Керування доступом");

    private final String label;

    AccessLevel(String label) {
        this.label = label;
    }

    /** Ukrainian name shown in the administration UI. */
    public String label() {
        return label;
    }

    /** Does holding this level satisfy a requirement of {@code required}? */
    public boolean allows(AccessLevel required) {
        return this.ordinal() >= required.ordinal();
    }

    /** The higher of two levels; either may be null, meaning "no access at all". */
    public static AccessLevel max(AccessLevel a, AccessLevel b) {
        if (a == null) return b;
        if (b == null) return a;
        return a.ordinal() >= b.ordinal() ? a : b;
    }

    /** Whether {@code held} (possibly null, meaning no grant) satisfies {@code required}. */
    public static boolean allows(AccessLevel held, AccessLevel required) {
        return held != null && held.allows(required);
    }

    /** Parses a level name, tolerating case and returning null for anything unrecognised. */
    public static AccessLevel parse(Object raw) {
        if (raw == null) return null;
        String name = raw.toString().trim().toUpperCase();
        for (AccessLevel level : values()) {
            if (level.name().equals(name)) return level;
        }
        return null;
    }
}
